### PR TITLE
[WIP] Fix RenderSliverFillViewport implementation

### LINKED ISSUE #23873: PageView with viewportFraction < 1.0 - GestureDetector is not working properly
Hey, looks like there is a bug preventing gesture detection on right side of the widget inside page view with only one element and viewport fraction smaller than 1.0.

This is how my widget tree looks like:
```
column
  - page view
    - gesture detector
      - green container
    - gesture detector
      - blue container
  - page view
    - gesture detector
      - red container
```

screenshot from 2018-11-02 16-14-59

Top gesture detectors (green&blue widgets) work fine, I can click anywhere and I can see debug logs in console.
Bottom gesture detector (red widget) works fine only for left half of the widget. When I click on the right half then onTap callback is not fired. Looks like something is preventing gesture detector from getting input.

flutter doctor output:
```
[✓] Flutter (Channel beta, v0.9.4, on Linux, locale en_US.UTF-8)
[✓] Android toolchain - develop for Android devices (Android SDK 27.0.3)
[✓] Android Studio (version 3.1)
[!] IntelliJ IDEA Community Edition (version 2018.2)
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
[✓] Connected devices (1 available)
```

Source code to reproduce bug:
```
import 'package:flutter/material.dart';

void main() => runApp(new MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return new MaterialApp(
      title: 'Flutter Demo',
      theme: new ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: Column(children: <Widget>[
        Expanded(
          child: ColorPages(
            colors: <Color>[
              Colors.green,
              Colors.blue,
            ],
          ),
        ),
        Expanded(
          child: ColorPages(
            colors: <Color>[
              Colors.red,
            ],
          ),
        ),
      ]),
    );
  }
}

class ColorPages extends StatefulWidget {
  final List<Color> colors;

  const ColorPages({Key key, this.colors}) : super(key: key);

  @override
  ColorPagesState createState() {
    return new ColorPagesState();
  }
}

class ColorPagesState extends State<ColorPages> {
  PageController _pageController;

  @override
  void initState() {
    super.initState();

    _pageController = new PageController(
      viewportFraction: 0.5,
    );
  }

  @override
  Widget build(BuildContext context) {
    return PageView.builder(
      controller: _pageController,
      itemBuilder: (context, index) {
        return GestureDetector(
          onTap: () {
            print(DateTime.now().toIso8601String() + " " + widget.colors[index].toString());
          },
          child: Container(
            color: widget.colors[index],
          ),
        );
      },
      itemCount: widget.colors.length,
    );
  }
}
```

### PR DESCRIPTION
**Pull request by LongCatIsLooong**
## Description

- Fixes the `PageView` hit testing issue by providing a more generic `RenderSliverFixedExtentBox.performLayout`. 
- Added missing docs to some overridden methods.

Currently in `RenderSliverFixedExtentBox.performLayout`'s implementation it assumes there's no padding around its children or gaps between its children. But such limitation is not stated in it's documentation. In fact some of its methods are documented 
>  ...... By default, returns a value consistent with the children being placed in order, without gaps, starting from layout offset zero.

that seems to imply we should be able to handle these variations easily by subclassing `RenderSliverFixedExtentBox` .

## Related Issues

Fixes #23873

## Tests

I added the following tests:

- All visible pages are able to receive touch events.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/rendering/sliver_fixed_extent_list.dart [hunk 1]
```diff
@@ -230,8 +235,8 @@ abstract class RenderSliverFixedExtentBoxAdaptor extends RenderSliverMultiBoxAda
     }
 
     final int lastIndex = indexOf(lastChild);
-    final double leadingScrollOffset = indexToLayoutOffset(itemExtent, firstIndex);
-    final double trailingScrollOffset = indexToLayoutOffset(itemExtent, lastIndex + 1);
+    final double leadingScrollOffset = firstIndex == 0 ? 0 : indexToLayoutOffset(itemExtent, firstIndex - 1) + itemExtent;
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/rendering/sliver_fixed_extent_list.dart]: Can I assume scroll offset always starts from 0?

File: packages/flutter/lib/src/rendering/sliver_fill.dart [hunk 2]
```diff
@@ -82,14 +118,14 @@ class RenderSliverFillViewport extends RenderSliverFixedExtentBoxAdaptor {
...
+    // Only overriding the documentation.
+    return super.estimateMaxScrollOffset(
       constraints,
       firstIndex: firstIndex,
       lastIndex: lastIndex,
-      leadingScrollOffset: leadingScrollOffset - padding,
-      trailingScrollOffset: trailingScrollOffset - padding,
-    ) + padding + padding;
+      leadingScrollOffset: leadingScrollOffset,
+      trailingScrollOffset: trailingScrollOffset,
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/sliver_fill.dart]: Why can we just ignore the padding?

File: packages/flutter/lib/src/rendering/sliver_fixed_extent_list.dart [hunk 3]
```diff
@@ -174,17 +180,16 @@ abstract class RenderSliverFixedExtentBoxAdaptor extends RenderSliverMultiBoxAda
...
-        final double max = computeMaxScrollOffset(constraints, itemExtent);
-        geometry = SliverGeometry(
-          scrollExtent: max,
-          maxPaintExtent: max,
-        );
-        childManager.didFinishLayout();
-        return;
-      }
+    if (firstChild == null &&
+      !addInitialChild(index: firstIndex, layoutOffset: indexToLayoutOffset(itemExtent, firstIndex))) {
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/sliver_fixed_extent_list.dart]: nit: indentation here is off.

File: packages/flutter/lib/src/rendering/sliver_fixed_extent_list.dart [hunk 4]
```diff
@@ -230,8 +235,8 @@ abstract class RenderSliverFixedExtentBoxAdaptor extends RenderSliverMultiBoxAda
     }
 
     final int lastIndex = indexOf(lastChild);
-    final double leadingScrollOffset = indexToLayoutOffset(itemExtent, firstIndex);
-    final double trailingScrollOffset = indexToLayoutOffset(itemExtent, lastIndex + 1);
+    final double leadingScrollOffset = firstIndex == 0 ? 0 : indexToLayoutOffset(itemExtent, firstIndex - 1) + itemExtent;
+    final double trailingScrollOffset = indexToLayoutOffset(itemExtent, lastIndex) + itemExtent;
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/sliver_fixed_extent_list.dart]: I don't understand the purpose of this change and the line above.

File: packages/flutter/lib/src/widgets/sliver.dart [hunk 5]
```diff
@@ -1057,6 +1057,24 @@ class SliverFillViewport extends SliverMultiBoxAdaptorWidget {
...
+    double leadingScrollOffset,
+    double trailingScrollOffset,
+  ) {
+    final double padding = (1 - viewportFraction) * constraints.viewportMainAxisExtent;
+    if (delegate.estimatedChildCount == null) {
+      return trailingScrollOffset + padding / 2;
+    }
+
+    return delegate.estimatedChildCount > 0
+      ? padding + viewportFraction * constraints.viewportMainAxisExtent * (delegate.estimatedChildCount)
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/sliver.dart]: remove the parentheses?

File: packages/flutter/lib/src/widgets/sliver.dart [hunk 6]
```diff
@@ -1057,6 +1057,24 @@ class SliverFillViewport extends SliverMultiBoxAdaptorWidget {
...
+  double estimateMaxScrollOffset(
+    SliverConstraints constraints,
+    int firstIndex,
+    int lastIndex,
+    double leadingScrollOffset,
+    double trailingScrollOffset,
+  ) {
+    final double padding = (1 - viewportFraction) * constraints.viewportMainAxisExtent;
+    if (delegate.estimatedChildCount == null) {
+      return trailingScrollOffset + padding / 2;
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/sliver.dart]: Why is this a good approximation?

File: packages/flutter/test/widgets/page_view_test.dart [hunk 7]
```diff
@@ -565,6 +565,52 @@ void main() {
...
+        );
+      }
+
+      visiblePages = const <int> [0, 1, 2];
+      await tester.pumpWidget(build());
+      // The first 3 items should be visible and tappable.
+
+      for (int index in visiblePages) {
+        expect(find.text(index.toString()), findsOneWidget);
+        // The center of 3 has an x-coordinate of 800, so we have to manually
```
Comment (goderbauer) [inline packages/flutter/test/widgets/page_view_test.dart]: What do you mean by this?

**Review by goderbauer — COMMENTED** [review]
I don't understand how this is fixing the issue. Can you add an explanation?

File: packages/flutter/lib/src/rendering/sliver_fill.dart [hunk 8]
```diff
@@ -57,23 +57,59 @@ class RenderSliverFillViewport extends RenderSliverFixedExtentBoxAdaptor {
     markNeedsLayout();
   }
 
+  // The padding that will be placed before and after the children if any, along the
+  // main axis.
+  //
+  // The default value is to center the earliest possible child or the last possible
```
[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/rendering/sliver_fill.dart]: Ah I publicized `padding` at some point and forgot to change the comment.

File: packages/flutter/lib/src/rendering/sliver_fill.dart (hunk 2 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/rendering/sliver_fill.dart]: The padding is now accounted for in `widget.estimateMaxScrollOffset` (in `packages/flutter/lib/src/widgets/sliver.dart`)

File: packages/flutter/lib/src/rendering/sliver_fill.dart [hunk 9]
```diff
@@ -82,14 +118,14 @@ class RenderSliverFillViewport extends RenderSliverFixedExtentBoxAdaptor {
     double leadingScrollOffset,
     double trailingScrollOffset,
   }) {
-    final double padding = _padding;
-    return childManager.estimateMaxScrollOffset(
+    // Only overriding the documentation.
```
[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/rendering/sliver_fill.dart]: The original documentation is no longer true. It used to say:
>   /// Must return the total distance from the start of the child with the
  /// earliest possible index to the end of the child with the last possible
  /// index.

which is not true now that we have a non-zero padding wrapping the children.

**Comment by LongCatIsLooong** [comment]
@goderbauer thanks for the review!

The paintExtent below doesn't look right when viewportFraction < 1. With this change paintExtent should be extended to cover the width of the screen when the current page is the first page.

paintExtent

 `PageView`'s render object `RenderSliverFillViewport` inherits from `RenderSliverFixedExtentBoxAdaptor`, but does not override its `performLayout` method. `RenderSliverFixedExtentBoxAdaptor.performLayout`'s implementation assumes there's no gaps between its children or padding before/after its children, that's not true in the case of `RenderSliverFillViewport` which does have padding. 

The changes made to `performLayout` gets rid of the said assumptions. If `firstIndex` is 0, the leading offset should start from 0, instead of the value of `_padding` (that's what `indexToLayoutOffset(itemExtent, 0)` returns).

File: packages/flutter/lib/src/widgets/sliver.dart (hunk 6 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/sliver.dart]: Without `delegate.estimatedChildCount` it's hard to estimate the max scroll offset. Should I just return `double.infinity` or `null`?

Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/sliver.dart]: Changed to `double.infinity`

File: packages/flutter/test/widgets/page_view_test.dart (hunk 7 above)
Comment (LongCatIsLooong) [reply packages/flutter/test/widgets/page_view_test.dart]: Does sound ambiguous after re-reading the comment. Updated.

**Comment by vitor-gyant** [comment]
Looking forward for having this merge into master. Thanks guys!

File: packages/flutter/lib/src/rendering/sliver_fixed_extent_list.dart (hunk 4 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/rendering/sliver_fixed_extent_list.dart]: Add trailing padding back to trailingScrollOffset, it should make more sense now.

File: packages/flutter/lib/src/rendering/sliver_fixed_extent_list.dart [hunk 10]
```diff
@@ -230,7 +235,7 @@ abstract class RenderSliverFixedExtentBoxAdaptor extends RenderSliverMultiBoxAda
     }
 
     final int lastIndex = indexOf(lastChild);
-    final double leadingScrollOffset = indexToLayoutOffset(itemExtent, firstIndex);
+    final double leadingScrollOffset = firstIndex == 0 ? 0 : indexToLayoutOffset(itemExtent, firstIndex - 1) + itemExtent;
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/sliver_fixed_extent_list.dart]: Not sure I understand the intentions here:

RenderSliverFillViewport.indexToLayoutOffset is:
```dart
return _padding + super.indexToLayoutOffset(itemExtent, index);
```
RenderSliverFixedExtendBoxAdaptor.indexToLayoutOffset (the superclass) is:
```dart
return itemExtent * index
```

So, ultimately, for end of this line becomes:
```
  indexToLayoutOffset(itemExtent, firstIndex - 1) + itemExtent
= _padding + super.indexToLayoutOffset(itemExtent, index - 1) + itemExtent
= _padding + (itemExtent * index - 1) + itemExtent
= _padding + itemExtend * index
= _padding + super.indexToLayoutOffset(itemExtent, index)
= indexToLayoutOffset(itemExtent, firstIndex)
```

So why do we need to subtract one from the index and then add another itemExtent?

Comment (goderbauer) [inline packages/flutter/lib/src/rendering/sliver_fixed_extent_list.dart]: The `firstIndex == 0 ? 0` also feels strange. It seems to half-encode into this class that there may be a padding proceeding the first element of the list, but nothing else in this class is doing that and it is instead the task of a subclass to define any kind of padding.

**Review by goderbauer — COMMENTED** [review]
I wonder if a simpler fix for this problem could be to re-implement SliverFillViewport by wrapping a regular SliverFixedExtentList in a SliverPadding via composition. That way, the SliverPadding would take care of all the oddities that are introduced by the padding (and we wouldn't have to implement (part of) the passing logic in RenderSliverFillViewport).

File: packages/flutter/lib/src/rendering/sliver_fixed_extent_list.dart (hunk 10 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/rendering/sliver_fixed_extent_list.dart]: `indexToLayoutOffset(itemExtent, firstIndex - 1) + itemExtent` is more generic. It remains valid if there're gaps between items.

**Comment by LongCatIsLooong** [comment]
Discussed this offline and we decided to take a more proper approach:
- Introduce a new widget `SliverLayoutBuilder`
- In `PageView`'s implementation, replace `RenderSliverFillViewport` with the new sliver layout builder 
- Deprecate `RenderSliverFillViewport`, announce the breakage.

I'm going to keep the PR open until the new PR arrives. 

@vitor-gyant sorry for the delay, unfortunately it's going to take a bit longer for a proper fix to land. If this is a blocking issue for you, a workaround would be applying the content of this PR to your local flutter sdk.

**Comment by vitor-gyant** [comment]
@LongCatIsLooong Thanks for letting me know. I'll wait for the official PR. Keep up the good work.

**Comment by goderbauer** [comment]
(PR triage): I'd suggest documenting the next steps in the associated issue and closing this PR since the PR in its current form is no longer actionable.

**Comment by LongCatIsLooong** [comment]
Close in favor of #35927

**Pull request closed without merging**